### PR TITLE
Improve reading/writing performance

### PR DESCRIPTION
**Pull request by RReverser**
Used various tricks to improve generic runtime performance.

Results on my machine as follows:

- before the change

```
decode vector tile with pbf x 858 ops/sec ±1.17% (90 runs sampled)
encode vector tile with pbf x 556 ops/sec ±1.46% (84 runs sampled)
```

- after the change

```
decode vector tile with pbf x 999 ops/sec ±1.13% (92 runs sampled)
encode vector tile with pbf x 638 ops/sec ±1.48% (84 runs sampled)
```

(both show ~15% improvement)

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -337,17 +306,7 @@ Pbf.prototype = {
...
-            len <= 0x3fff ? 2 :
-            len <= 0x1fffff ? 3 :
-            len <= 0xfffffff ? 4 : Math.ceil(Math.log(len) / (Math.LN2 * 7));
-
-        // if 1 byte isn't enough for encoding message length, shift the data to the right
-        if (varintLen > 1) {
-            this.realloc(varintLen - 1);
-            for (var i = this.pos - 1; i >= startPos; i--) this.buf[i + varintLen - 1] = this.buf[i];
-        }
+        reallocForRawMessage(startPos, len, this);
```
Comment (mourner) [inline index.js]: Is there a benefit to separating this one? The body doesn't do a lot and the separated function is called every time...

File: index.js [hunk 2]
```diff
@@ -415,6 +374,45 @@ Pbf.prototype = {
...
+
+    var maxPos = pbf.pos + 10;
+
+    while (val >= 1) {
+        if (pbf.pos >= maxPos) throw new Error('Given varint doesn\'t fit into 10 bytes');
+        var b = val & 0xff;
+        pbf.buf[pbf.pos++] = b | (val >= 0x80 ? 0x80 : 0);
+        val /= 0x80;
+    }
+}
```
Comment (mourner) [inline index.js]: Let's use consistent names — e.g. `readBigVarint` and `writeBigVarint`.

**Comment by mourner** [comment]
Brilliant! Will be happy to merge this one.

File: index.js (hunk 1 above)
Comment (RReverser) [inline index.js]: Yes, that part is important for performance (as well as in other places). It doesn't work as straightforward as you think :)

Comment (RReverser) [inline index.js]: (you can try to revert the change locally and observe the difference)

File: index.js (hunk 2 above)
Comment (RReverser) [inline index.js]: Sure, I can change that. I just thought that it makes sense to call them differently because first helper reads only extra "digits" for the existing number, while the second one writes big number as whole. What do you think?

Comment (mourner) [inline index.js]: Ah, yeah, you're right. Maybe `readVarintRemainder` and `writeBigVarint`? The word "extension" feels a bit weird to me in this context.

Comment (RReverser) [inline index.js]: Sure thing.

Comment (RReverser) [inline index.js]: Done.

File: index.js (hunk 1 above)
Comment (mourner) [inline index.js]: I can see how cutting out less frequently taken code path in other places helps with V8 function inlining, but I don't understand why it helps in this particular case. Can you explain?

Comment (RReverser) [inline index.js]: Probably you're right, it's not very obvious in this particular case and thus not very future-proof for the maintainance. Moved the condition out to simplify it.

Comment (mourner) [inline index.js]: Another thing I don't like is that the name says `realloc`, but it's not only reallocating but also shifting the encoded data.
